### PR TITLE
Add an example UserUpdateHistory class and SQL

### PR DESCRIPTION
**Pull request by gbrodman**
This won't pass the build because it doesn't follow the necessary logic of adding the SQL code to the golden file first, but it serves as an example of what the code for console update history objects will look like.

Notably:
- We split User, adding AbstractUser, to store an AbstractUser in UserUpdateHistory (analogous to DomainBase/Domain et al.)
- We have an abstract ConsoleUpdateHistory class that all histories will extend from (analogous to HistoryEntry)

RegistrarUpdateHistory and RegistrarPocUpdateHistory will look similar,
but ConsoleEppActionHistory will only contain the linked history class
type and ID

- - -
This change is [Reviewable](https://reviewable.io/reviews/google/nomulus/2330)

### CODE REVIEW

File: core/src/main/java/google/registry/model/console/AbstractUser.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * A console user, either a registry employee or a registrar partner.
+ *
+ * <p>This class deliberately does not include an {@link Id} so that any foreign-keyed fields can
+ * refer to the proper parent entity's ID, whether we're storing this in the DB itself or as part of
+ * another entity.
+ */
+@Access(AccessType.FIELD)
+@Embeddable
+@MappedSuperclass
+public class AbstractUser extends UpdateAutoTimestampEntity implements Buildable {
```
[the remark this answers is not in the document]
Comment (gbrodman) [reply core/src/main/java/google/registry/model/console/AbstractUser.java:46]: this is spurious; we do this everywhere

File: core/src/main/java/google/registry/model/console/UserUpdateHistory.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * A persisted history object representing an update to a User.
+ *
+ * <p>In addition to the generic history fields (time, URL, etc.) we also persist a copy of the
+ * modified User object at this point in time.
+ */
+@Access(AccessType.FIELD)
+@Entity
+@Table(indexes = {@Index(columnList = "historyActingUser")})
+public class UserUpdateHistory extends ConsoleUpdateHistory {
```
[the remark this answers is not in the document]
Comment (gbrodman) [reply core/src/main/java/google/registry/model/console/UserUpdateHistory.java:35]: this is spurious; we do this everywhere

**Review by jianglai — CHANGES_REQUESTED** [review]
> Reviewed 3 of 6 files at r1, 4 of 9 files at r2, 4 of 4 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2330)* status: all files reviewed, 7 unresolved discussions (waiting on @gbrodman and @github-advanced-security[bot])

___
*[`config/presubmits.py` line 94 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLWZIAAQ1gjvPy94br:-NrLWZIAAQ1gjvPy94bs:bq92iv0) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/config/presubmits.py#L94)):*
> ```Python
>             "registrar_dbg.", "google-java-format-diff.py",
>             "nomulus.golden.sql", "soyutils_usegoog.js", "javascript/checks.js",
>             "/src/main/generated", "/src/test/generated"
> ```

Interesting, why was it not a problem before...?
___
*[`config/presubmits.py` line 288 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLVabx9cuBRpK_EBie:-NrLVabx9cuBRpK_EBif:bmxnmfr) ([raw file](https://github.com/google/nomulus/blob/abc1a0ef3d1e146588536d8454b73f352b70b967/config/presubmits.py#L288)):*
> ```Python
> 
> if __name__ == "__main__":
>   print('python version is %s' % sys.version)
> ```

It might be useful to keep this output for debugging purposes.
___
*[`core/src/main/java/google/registry/model/console/AbstractUser.java` line 46 at r1](https://reviewable.io/reviews/google/nomulus/2330#gh-1496442212:-NrLXjr0BZsYzhgPEY9B:b114wyz) ([raw file](https://github.com/google/nomulus/blob/1ea9173ece4b76efd678cc5ba2cf63f14ec612dc/core/src/main/java/google/registry/model/console/AbstractUser.java#L46)):*
<details><summary><i>Previously, gbrodman wrote…</i></summary><blockquote>

this is spurious; we do this everywhere
</blockquote></details>

Is this class just renamed from `User` with some of minor changes
___
*[`core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java` line 74 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLZVOX2DmUAzedmzrJ:-NrLZVOX2DmUAzedmzrK:b3nsk9y) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java#L74)):*
> ```Java
>   /** The type of modification. */
>   @Column(nullable = false, name = "historyType")
>   // TODO make an enum
> ```

I think it is possible to change a column type from String to Enum after the fact. But we probably should just define the enums from the get go to avoid the messy conversion.
___
*[`core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java` line 113 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrL_CdIDEcHEBklTrZx:-NrL_CdIDEcHEBklTrZy:bd65ecx) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java#L113)):*
> ```Java
>   /** Builder for the immutable ConsoleUpdateHistory. */
>   public abstract static class Builder<
>           T extends ConsoleUpdateHistory, B extends ConsoleUpdateHistory.Builder<?, ?>>
> ```

Did I mention I hate Java generics... It's like a brain teaser to unwrap this.
___
*[`core/src/main/java/google/registry/model/console/UserUpdateHistory.java` line 42 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLcfcT6ieVB0pD62lt:-NrLcfcT6ieVB0pD62lu:bfhmvcn) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/core/src/main/java/google/registry/model/console/UserUpdateHistory.java#L42)):*
> ```Java
> 
>   // This field exists so that it's populated in the SQL table
>   @Column(nullable = false)
> ```

Do you think the column is better to be called "id" (so it uniform across all console history entry times), or more precisely as "user_id"?

**Review by gbrodman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2330)* status: all files reviewed, 7 unresolved discussions (waiting on @github-advanced-security[bot] and @jianglai)

___
*[`config/presubmits.py` line 94 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLWZIAAQ1gjvPy94br:-NrLpS3E2qHGPO1vJtlz:b-vkpb8q) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/config/presubmits.py#L94)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

Interesting, why was it not a problem before...?
</blockquote></details>

I'm not sure...
___
*[`config/presubmits.py` line 288 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLVabx9cuBRpK_EBie:-NrLphgi3LYacAcFWwFj:b-896fix) ([raw file](https://github.com/google/nomulus/blob/abc1a0ef3d1e146588536d8454b73f352b70b967/config/presubmits.py#L288)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

It might be useful to keep this output for debugging purposes.
</blockquote></details>

Done.
___
*[`core/src/main/java/google/registry/model/console/AbstractUser.java` line 46 at r1](https://reviewable.io/reviews/google/nomulus/2330#gh-1496442212:-NrLpPZp-q8Ba2P_cFdA:bb7s9mg) ([raw file](https://github.com/google/nomulus/blob/1ea9173ece4b76efd678cc5ba2cf63f14ec612dc/core/src/main/java/google/registry/model/console/AbstractUser.java#L46)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

Is this class just renamed from `User` with some of minor changes
</blockquote></details>

yeah
___
*[`core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java` line 74 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLZVOX2DmUAzedmzrJ:-NrLpkAm5SjXxqMoKwZ_:b-5c0x76) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java#L74)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

I think it is possible to change a column type from String to Enum after the fact. But we probably should just define the enums from the get go to avoid the messy conversion.
</blockquote></details>

So, for the SQL that we'll eventually merge this is immaterial because it's going to be a `text` type either way. I haven't defined the enum yet because 1. we need the SQL first anyway 2. it's not entirely clear precisely what values the enum will have. I suppose we can always add values on as we figure it out though. 
___
*[`core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java` line 113 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrL_CdIDEcHEBklTrZx:-NrLmGps58sz8gW7xrza:b-u8z4ku) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java#L113)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

Did I mention I hate Java generics... It's like a brain teaser to unwrap this.
</blockquote></details>

oh man yeah it's uh, rough 
___
*[`core/src/main/java/google/registry/model/console/UserUpdateHistory.java` line 42 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLcfcT6ieVB0pD62lt:-NrLnosBDqIoXIGSkmsO:b-fscj43) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/core/src/main/java/google/registry/model/console/UserUpdateHistory.java#L42)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

Do you think the column is better to be called "id" (so it uniform across all console history entry times), or more precisely as "user_id"?
</blockquote></details>

yeah might be clearer to use user_id

**Review by jianglai — APPROVED** [review]
> Reviewed 7 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2330)* status: all files reviewed, 3 unresolved discussions (waiting on @gbrodman and @github-advanced-security[bot])

___
*[`core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java` line 74 at r4](https://reviewable.io/reviews/google/nomulus/2330#-NrLZVOX2DmUAzedmzrJ:-NrfAMOQ-DJO6NHeDkeb:b-invipm) ([raw file](https://github.com/google/nomulus/blob/340a7661860912407362fc13e43bc0d92beec86e/core/src/main/java/google/registry/model/console/ConsoleUpdateHistory.java#L74)):*
<details><summary><i>Previously, gbrodman wrote…</i></summary><blockquote>

So, for the SQL that we'll eventually merge this is immaterial because it's going to be a `text` type either way. I haven't defined the enum yet because 1. we need the SQL first anyway 2. it's not entirely clear precisely what values the enum will have. I suppose we can always add values on as we figure it out though.
</blockquote></details>

It looks like with PostgreSQL 12 you can trivially add a new value to an enum column, so you don't have to define all of them before you create the column:

https://stackoverflow.com/questions/1771543/adding-a-new-value-to-an-existing-enum-type

But we are on PostgreSQL 11, so...

**Pull request closed without merging**